common/comms/src/inbound_message_service/msg_processing_worker.rs

```diff
@@ -0,0 +1,237 @@
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::inbound_message_service::message_dispatcher::{Dispatchable, MessageDispatcher};
+use std::{convert::TryFrom, marker::Send, thread};
+use tari_comms::connection::{
+    message::FrameSet,
+    types::SocketType,
+    zmq::{Context, InprocAddress, ZmqEndpoint},
+};
+use tari_crypto::keys::PublicKey;
+
+pub struct MsgProcessingWorker<PubKey, DispMsg: Dispatchable>
+where PubKey: PublicKey
+{
+    context: Context,
+    inbound_address: InprocAddress,
+    node_identity: PubKey,
+    message_dispatcher: MessageDispatcher<DispMsg>,
+}
+
+impl<PubKey: PublicKey + Send + 'static, DispMsg: Dispatchable + TryFrom<FrameSet> + 'static>
+    MsgProcessingWorker<PubKey, DispMsg>
+{
+    /// Setup a new MsgProcessingWorker that will read incoming messages and dispatch them using the message_dispatcher
+    pub fn new(
+        context: Context,
+        inbound_address: InprocAddress,
+        node_identity: PubKey,
+        message_dispatcher: MessageDispatcher<DispMsg>,
+    ) -> MsgProcessingWorker<PubKey, DispMsg>
+    {
+        MsgProcessingWorker {
+            context,
+            inbound_address,
+            node_identity,
+            message_dispatcher,
+        }
+    }
+
+    /// Start the MsgProcessingWorker thread, connect to reply socket, retrieve and dispatch incoming messages to
+    /// handlers
+    pub fn start(self) {
+        thread::spawn(move || {
+            loop {
+                // Establish connection inbound socket
+                match self.context.socket(SocketType::Reply) {
```

Comment:
`and_then()` FTW

Reply:
Reduced the nested match statements significantly